File: linked-list/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {

        // l3 initial head is 0(dummy)
        //l1 and l2 
        // value = l1(v) + l2(v)
        // carry = value/10 - 
        // l3(next) = value%10 - 
        // all nodes in list becomes  equal to their next
        // l1 or l2
        // value = l1(v) or l2(v)
        // carry = value/10
        //l3(v) = value % 10 
        //if carry -> l3(v) = carry
        // return l3 head(0 dummy) next

        ListNode* l3 = new ListNode(0);
        int value;
        int carry = 0;
        ListNode* head = l3; 

        while(l1 && l2)
        {
            value = l1->val + l2->val + carry;
            carry = value/10;
            l3->next = new ListNode(value%10);
            l1 = l1->next;
            l2 = l2-> next;
            l3 = l3->next;
        }
        while(l1)
        {
            value = l1->val + carry;
            carry = value/10;
            l3->next = new ListNode(value%10);
            l1 = l1->next;
            l3 = l3->next;
        }
        while(l2)
        {
            value = l2->val + carry;
            carry = value/10;
            l3->next = new ListNode(value%10);
            l2 = l2->next;
            l3 = l3->next;
        }

        if(carry)
        {
            l3->next = new ListNode(carry);
            l3 = l3->next;
        }

        return head-> next;

        
    }
};

File: recursion/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        # non empty non neg linked list -> digits in reverse order
        # sum them and return linked list 
        # example - [2,4,3], [5,6,4] = [7,0,8] 342 + 456
        # need to know how you sum numbers -> sum them or take carry 
        # interate through both nodes and add them to new linked list node 
        # if res greater than 9, add carry to next node
        # if linked list are of diff lenght, assume it's a zero for all remaining nodes where they don't have node from 2nd linked list 
        # if you have a carry, but no remaining nodes, create a node for the carry 

        #new linked list 
        dummy = ListNode()
        curr = dummy 

        #carry for addition 
        carry = 0

        # until either linked list is empty, or if it has carry but no node remaining, create another node just for carry 
        while l1 or l2 or carry:

            #values of both lists 
            v1 = l1.val if l1 else 0 # if l1 is null put 0 as its value 
            v2 = l2.val if l2 else 0 

            val = v1 + v2 + carry 

            #what if this value is larger than 10 
            carry = val // 10 
            val = val%10

            #update pointers
            curr = curr.next
            l1 = l1.next if l1 else None 
            l2 = l2.next if l2 else None 
        
        return dummy.next